#!/usr/bin/env python
# Configuration for virtual display
# Based on tutorial at http://www.ros.org/wiki/dynamic_reconfigure/Tutorials/SettingUpDynamicReconfigureForANode

PACKAGE='camera_aravis'
import roslib; roslib.load_manifest(PACKAGE)

from dynamic_reconfigure.parameter_generator import *
from driver_base.msg import SensorLevels

gen = ParameterGenerator()

arvauto_enum = gen.enum([gen.const("Off",        int_t, 0, "Use Manual Settings"),
                         gen.const("Once",       int_t, 1, "Recalc Once"),
                         gen.const("Continuous", int_t, 2, "Recalc Continually") ],
                         "Automatic Settings")
arvmode_enum = gen.enum([gen.const("FreeRunning", int_t, 0, "Continuous Internal Triggering"),
                         gen.const("SingleFrame", int_t, 1, "Capture One Frame on Trigger")],
                         "Acquisition Mode")
trigger_enum = gen.enum([gen.const("Software",   int_t, 0, "Triggered via software"),
                         gen.const("Line1",      int_t, 1, "Hardware input 1"),
                         gen.const("Line2",      int_t, 2, "Hardware input 2") ],
                         "Trigger Source")

gen.add("autoexposure",    int_t,    SensorLevels.RECONFIGURE_RUNNING, "Automatic exposure", 0, 0, 2, edit_method=arvauto_enum)
gen.add("autogain",        int_t,    SensorLevels.RECONFIGURE_RUNNING, "Automatic gain",     0, 0, 2, edit_method=arvauto_enum)
gen.add("exposure",        double_t, SensorLevels.RECONFIGURE_RUNNING, "Exposure time (us)", 2000.0, 0, 1e6)
gen.add("gain",            int_t,    SensorLevels.RECONFIGURE_RUNNING, "Gain (%)",           100, 0, 10000)
gen.add("acquisitionmode", int_t,    SensorLevels.RECONFIGURE_RUNNING, "Acquisition Mode",   0, 0, 1, edit_method=arvmode_enum)
gen.add("framerate",       double_t, SensorLevels.RECONFIGURE_RUNNING, "Framerate (fps)",    100.0, 0.0, 1000.0)
gen.add("triggersource",   int_t,    SensorLevels.RECONFIGURE_RUNNING, "Trigger Source",     0, 0, 2, edit_method=trigger_enum)
gen.add("softwarerate",    double_t, SensorLevels.RECONFIGURE_RUNNING, "Software Trigger Rate (hz)", 100.0, 0.01, 200.0)
#gen.add("xRoi",           int_t,    SensorLevels.RECONFIGURE_RUNNING, "ROI x offset",       0, 0, 4096)
#gen.add("yRoi",           int_t,    SensorLevels.RECONFIGURE_RUNNING, "ROI y offset",       0, 0, 4096)
#gen.add("widthRoi",       int_t,    SensorLevels.RECONFIGURE_RUNNING, "ROI width",          -1, -1, 4096)
#gen.add("heightRoi",      int_t,    SensorLevels.RECONFIGURE_RUNNING, "ROI height",         -1, -1, 4096)
gen.add("frame_id",        str_t,    SensorLevels.RECONFIGURE_RUNNING, "ROS camera frame",   "camera")
gen.add("focuspos",        int_t,    SensorLevels.RECONFIGURE_RUNNING, "FocusPos",           32767, 0, 65535)

exit(gen.generate(PACKAGE, "camera_aravis_params", "CameraAravis"))
